Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,263 @@
+                self.pollClosure?()
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    self.stateMachineLock.withLockedValue { $0.shutDown() }
+                    break actionLoop
+                }
+            case .suspendPollLoop:
+                await withTaskCancellationHandler {
+                    await withCheckedContinuation { continuation in
+                        self.stateMachineLock.withLockedValue { $0.suspendLoop(continuation: continuation) }
+                    }
+                } onCancel: {
+                    self.stateMachineLock.withLockedValue { $0.shutDown() }
+                }
+            case .shutdownPollLoop:
+                self.stateMachineLock.withLockedValue { $0.shutDown() }
+                break actionLoop
+            }
+        }
+    }
+
+    /// The delivery report callback function that handles acknowledged messages.
+    private(set) lazy var deliveryReportCallback: (UnsafePointer<rd_kafka_message_t>?) -> Void = { messagePointer in
+        guard let messagePointer = messagePointer else {
+            self.logger.error("Could not resolve acknowledged message")
+            return
+        }
+
+        let messageID = UInt(bitPattern: messagePointer.pointee._private)
+
+        let messageResult: Element
+        do {
+            let message = try KafkaAcknowledgedMessage(messagePointer: messagePointer, id: messageID)
+            messageResult = .success(message)
+        } catch {
+            guard let error = error as? KafkaAcknowledgedMessageError else {
+                fatalError("Caught error that is not of type \(KafkaAcknowledgedMessageError.self)")
+            }
+            messageResult = .failure(error)
+        }
+
+        self.stateMachineLock.withLockedValue { stateMachine in
+            let yieldResult = stateMachine.yield(messageResult)
+            switch yieldResult {
+            case .produceMore:
+                stateMachine.produceMore()
+            case .stopProducing:
+                stateMachine.stopProducing()
+            case .dropped:
+                stateMachine.shutDown()
+            }
+        }
+        // The messagePointer is automatically destroyed by librdkafka
+        // For safety reasons, we only use it inside of this closure
+    }
+}
+
+extension KafkaBackPressurePollingSystem: NIOAsyncSequenceProducerDelegate {
+    func produceMore() {
+        self.stateMachineLock.withLockedValue { $0.produceMore() }
+    }
+
+    func didTerminate() {
+        self.stateMachineLock.withLockedValue { $0.shutDown() }
+    }
+}
+
+extension KafkaBackPressurePollingSystem {
+    /// The state machine used by the ``KafkaBackPressurePollingSystem``.
+    struct StateMachine: Sendable {
+        /// The ``NIOAsyncSequenceProducer.Source`` used for yielding the messages to the ``NIOAsyncSequenceProducer``.
+        var sequenceSource: Producer.Source? // TODO: make sendable?
```

Comment:
[`NIOAsyncSequenceProducer.Source`](https://github.com/apple/swift-nio/blob/6213ba7a06febe8fef60563a4a7d26a4085783cf/Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift#L211) is not `Sendable` at the moment which is giving us a warning here. Should I try making it `Sendable` in `swift-nio` or do we have a workaround for that in our case here?

Reply:
The `StateMachine` doesn't necessarily need to be `Sendable` we are holding it in a `Lock` which makes it `Sendable` We just need to use `@unchecked Sendable` on the `KafkaBackPressurePollingSystem`